Handle gaps in estimate data in LTMPlot #2534

riccioclista opened the pull request:
This PR makes the plotting of model estimate data more robust against gaps in the estimate data. Normally estimate data (CP, W' Prime, ...) is stored sequentially without gaps (each point in time from a defined start to a defined end date has a model estimate). However, if this is for some reason (bug?) not the case, LTMPlot.cpp would still plot the data without leaving the proper gap (See #2534). This PR inserts 0s in LTMPlot for time spans without estimate data.

liversedge commented:
What is the problem this is trying to fix?

Gaps in data are perfectly valid -- the plot should also contain gaps.

Is there an issue when aggregating (by week/month etc) ?

riccioclista commented:
The problem are gaps in the power model estimate data (`context->athlete->getPDEstimates()`). Model estimate entries in this list have a start and end date. For LTMPlot to work correctly, this list must not contain any time gaps, because LTMPlot ignores them and just plots all data in sequence. This can be observed on the "Trends/CP History" chart with by-day resolution.

riccioclista commented:
Here is the CP history chart with current master:
![screenshot from 2017-10-28 15-30-21](https://user-images.githubusercontent.com/759436/32135080-47a99ddc-bbf9-11e7-9591-e740983c74f2.png)
and here with fix:
![screenshot from 2017-10-28 15-27-47](https://user-images.githubusercontent.com/759436/32135084-53bbd572-bbf9-11e7-8d64-59df781bd3c0.png)

In the fixed version, you can see that gaps exist and are accounted for.

liversedge commented:
Not sure I prefer the fixed version, and need to check how data is aggregated when show month / year...

There is a special plot curve called QwtPlotGappedCurve that may also be useful to show gaps (instead of dropping to zero) perhaps that is a better compromise ?

riccioclista commented:
QwtPlotGappedCurve might be a better solution, I'll look into it.

liversedge commented:
Cool, you will see it is used to plot power in AllPlot.cpp, if you want an example of it in use.
CHEERS, Mark.

riccioclista commented:
The CP history chart now looks like this:
![screenshot from 2017-10-29 10-53-12](https://user-images.githubusercontent.com/759436/32142436-6432bd3c-bc97-11e7-941d-c13e7d85f476.png)

liversedge commented:
Excellent, have you checked it works correctly (and calculates correctly) when you show by week/month/year ? (i.e. aggregating)

liversedge commented:
Also, I noticed you changed qwtplotgappedcurve, please remove that change. It is used in other places.

riccioclista commented:
I've removed the QwtPlotGappedCurve change. However, there are 2 problems that come with the switch to QwtPlotGappedCurve:

* QwtPlotGappedCurve omits the last element in the data series. The upper (exclusive) bound is determind like so: `to = dataSize() - 1;`. This can be best observed when aggregating by month:
![screenshot from 2017-10-31 11-34-46](https://user-images.githubusercontent.com/759436/32220117-d6c187bc-be30-11e7-86a3-4c61ad8c75f8.png)

* QwtPlotGappedCurve connects the last data point of a gap with the first data point after the gap resulting in a very steep line:
![screenshot from 2017-10-31 10-33-36](https://user-images.githubusercontent.com/759436/32220253-5b3e9732-be31-11e7-8aa1-4771f85c83d2.png)

Other than that, I've not observed any problems.

liversedge commented:
Thanks ! We can fix the fencepost error separately.
Can you confirm the aggregate is not affected by zero values?

If so, I'll merge, thank you for being patient with me :)

riccioclista commented:
No, it's not good to merge yet. You're right, aggregates might not get calculated correctly.

riccioclista commented:
I've fixed it now for all aggregate options. There were 2 issues in the previous version that I've fixed now:
* Before performing a gap fill-up in an aggregate chart, the last item on the aggregate stack (yTotal/xCount) is flushed, i.e. the avg is calculated and written to the values array `y`.
* The calculation of the previously processed estimate should only use the estimate end date if the chart resolution is by-day. For any other chart aggregate type the start date should be used, because the end date might already be in a subsequent aggregate group.

I'm sorry, I didn't spot the issues earlier. I've now tested all aggregate options thoroughly by comparing the calculated values with the model values in a [spreadsheet](https://github.com/GoldenCheetah/GoldenCheetah/files/1434205/testdata.xlsx).

liversedge commented:
Excellent. Will merge then we can test via the development builds.
